### ISSUE TITLE
In this situation, is it possible to render without a camera view?

### ISSUE THREAD

**Issue by kcw4875** [facebookresearch/pytorch3d#130]
## ❓ Questions on how to use PyTorch3D

My situation is as follows.
When the vertices are already rotated and projected and 
if the desired render image size is 224x224, the vertices are normalized to be positioned at 224 sizes.

In my opinion, I just need to plot using the given vertice's x, y coordinates, and color information and triangle without considering the camera view(angle and position..).

###############
input  
vertices : [x, 3] (tensor), (rotated, projected, and normalized(224x224))
triangle : [y, 3] (tensor) , color = [x, 3] (tensor)

output
render image (224x224)
########################
is this possible in pytorch3d?. If possible, can you tell me how?
plz..

**Comment by nikhilaravi** [comment]
@kcw4875  You will first need to create a `Meshes` with the verts/faces of the mesh and a `Textures` object with the vertex_rgb colors.  See this related issue about how to render a mesh with vertex colors: #112 and have a look at the [camera position optimization tutorial](https://github.com/facebookresearch/pytorch3d/blob/master/docs/tutorials/camera_position_optimization_with_differentiable_rendering.ipynb) which shows how to initialize a Mesh with texture, initialize a renderer, and how to specify the settings such as image_size.  

If the vertices are already projected into 2D you could write your own rasterizer which avoids the camera transformation step. So instead of the `MeshRasterizer` in the camera position optimization tutorial you would replace it with the `CustomMeshRasterizer` below and omit passing in the `cameras` during initialization.

```python
class CustomMeshRasterizer(nn.Module):
    """
    This class implements methods for rasterizing a batch of heterogenous
    Meshes which have already been transformed.
    """

    def __init__(self, cameras, raster_settings=None):
        """
        Args:
            raster_settings: the parameters for rasterization. This should be a
                named tuple.
        """
        super().__init__()
        if raster_settings is None:
            raster_settings = RasterizationSettings()
        self.raster_settings = raster_settings

    def forward(self, meshes_world, **kwargs) -> Fragments:
        """
        Args:
            meshes_screen: a Meshes object representing a batch of meshes with
                          coordinates already projected.
        Returns:
            Fragments: Rasterization outputs as a named tuple.
        """
        raster_settings = kwargs.get("raster_settings", self.raster_settings)
        pix_to_face, zbuf, bary_coords, dists = rasterize_meshes(
            meshes_screen,
            image_size=raster_settings.image_size,
            blur_radius=raster_settings.blur_radius,
            faces_per_pixel=raster_settings.faces_per_pixel,
            bin_size=raster_settings.bin_size,
            max_faces_per_bin=raster_settings.max_faces_per_bin,
            perspective_correct=raster_settings.perspective_correct,
        )
        return Fragments(
            pix_to_face=pix_to_face, zbuf=zbuf, bary_coords=bary_coords, dists=dists
        )
```

Do you only want the color of the closest face in the z direction? If so use the `HardPhongShader` class for the shader.

**Comment by kcw4875** [comment]
Thank you for the quick response,


The meshes_screen is missing in the forward part of the CustomMeshRasterizer code you gave.

I put "meshs_screen = mesh_world" and run.
However, all pixel values ​​in the rendered image are 0.


In the camera position optimization tutorial, I changed my code to fit the customMeshRasterizer.
The code is as follows.

**Comment by nikhilaravi** [comment]
Are you using the same teapot_mesh as in the tutorial? It will not work with the teapot mesh which is loaded because the teapot mesh vertices are in world coordinates and need to be transformed and projected into 2D. 

You mentioned that your mesh vertices have already been transformed and projected - did you try it with this input?

**Comment by nikhilaravi** [comment]
@kcw4875 do you still need help with this or can we close this issue?

**Comment by 0xsuu** [comment]
Hi @nikhilaravi,

When I was trying out rendering without a camera view on the latest version (0.2.0), I found that the vertex coordinates are required to be normalised to 0.0 ~ 1.0, any vertex with z-axis value less than 0 will not be rendered and the image is upside down. Rather than what is described in this issue. I applyed the CustomMeshRasterizer and SimpleShader in #84. Did updates caused it or did I miss something?

Also, the current HardPhongShader requires a camera argument and checks it, is there a proper way to bypass that?

Thanks in advance for any insights!
